src/utils/consistency/mod.rs

```diff
@@ -1,59 +1,300 @@
+        .context("Loading crate data from database for consistency check")?;
 
     tracing::info!("Loading data from index...");
-    let timer = std::time::Instant::now();
     let index_data =
-        self::index::load(index).context("Loading crate data from index for consistency check")?;
-    info!("...loaded in {:?}", timer.elapsed());
-
-    let diff = db_data.diff(index_data);
-
-    for krate in diff.crates {
-        match krate {
-            Diff::Both(name, diff) => {
-                for release in diff.releases {
-                    match release {
-                        Diff::Both(_, _) => {}
-                        Diff::Left(version, _) => {
-                            println!("Release in db not in index: {} {}", name, version);
-                        }
-                        Diff::Right(version, _) => {
-                            println!("Release in index not in db: {} {}", name, version);
-                        }
+        index::load(&index).context("Loading crate data from index for consistency check")?;
+
+    let diff = diff::calculcate_diff(db_data.iter(), index_data.iter());
+    let result = handle_diff(ctx, diff.iter(), dry_run)?;
+
+    println!("============");
+    println!("SUMMARY");
+    println!("============");
+    println!("difference found:");
+    for (key, count) in diff.iter().counts_by(|el| match el {
+        diff::Difference::CrateNotInIndex(_) => "CrateNotInIndex",
+        diff::Difference::CrateNotInDb(_, _) => "CrateNotInDb",
+        diff::Difference::ReleaseNotInIndex(_, _) => "ReleaseNotInIndex",
+        diff::Difference::ReleaseNotInDb(_, _) => "ReleaseNotInDb",
+        diff::Difference::ReleaseYank(_, _, _) => "ReleaseYank",
+    }) {
+        println!("{} => {}", key, count);
+    }
+
+    println!("============");
+    if dry_run {
+        println!("activities that would have been triggered:");
+    } else {
+        println!("activities triggered:");
+    }
+    println!("builds queued: {}", result.builds_queued);
+    println!("crates deleted: {}", result.crates_deleted);
+    println!("releases deleted: {}", result.releases_deleted);
+    println!("yanks corrected: {}", result.yanks_corrected);
+
+    Ok(())
+}
+
+#[derive(Default)]
+struct HandleResult {
+    builds_queued: u32,
+    crates_deleted: u32,
+    releases_deleted: u32,
+    yanks_corrected: u32,
+}
+
+fn handle_diff<'a, I>(ctx: &dyn Context, iter: I, dry_run: bool) -> Result<HandleResult>
+where
+    I: Iterator<Item = &'a diff::Difference>,
+{
+    let mut result = HandleResult::default();
+
+    let mut conn = ctx.pool()?.get()?;
+    let storage = ctx.storage()?;
+    let config = ctx.config()?;
+    let build_queue = ctx.build_queue()?;
+
+    for difference in iter {
+        println!("{difference}");
+
+        match difference {
+            diff::Difference::CrateNotInIndex(name) => {
+                if !dry_run {
+                    delete::delete_crate(&mut conn, &storage, &config, name)?;
```

Comment:
Should we log and continue on individual errors during this?

Reply:
valid piont, I updated the logic.